crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -285,42 +288,17 @@ impl Account {
 
         let mut decrypted: Option<(Session, String)> = None;
 
+        // Try to decrypt the message using each Session we share with the
+        // given curve25519 sender key.
         for session in &mut *sessions.lock().await {
-            let mut matches = false;
-
-            // If this is a pre-key message check if it was encrypted for our
-            // session, if it wasn't decryption will fail so no need to try.
-            if let OlmMessage::PreKey(m) = &message {
-                matches = session.matches(sender_key, m.clone()).await?;
-
-                if !matches {
-                    continue;
-                }
-            }
-
-            let ret = session.decrypt(message.clone()).await;
-
-            match ret {
-                Ok(p) => {
-                    decrypted = Some((session.clone(), p));
-                    break;
-                }
-                Err(e) => {
-                    // Decryption failed with a matching session, the session is
-                    // likely wedged and needs to be rotated.
-                    if matches {
-                        warn!(
-                            sender = sender.as_str(),
-                            sender_key = sender_key,
-                            error =? e,
-                            "Found a matching Olm session yet decryption failed",
-                        );
-                        return Err(OlmError::SessionWedged(
-                            sender.to_owned(),
-                            sender_key.to_owned(),
-                        ));
-                    }
-                }
+            if let Ok(p) = session.decrypt(message).await {
+                decrypted = Some((session.clone(), p));
+                break;
+            } else {
+                // An error here is completely normal, after all we don't know
+                // which session was used to encrypt a message. We will log a
+                // warning if no session was able to decrypt the message.
```

Comment:
```suggestion
                // These errors are expected to happen since we don't know which
                // session was used to encrypt a message, so we will undoubtedly
                // attempt to decrypt using the wrong one. In the event that no
                // session manages to decrypt the message, we will log a warning
                // in upstream code.
```

My attempt to make it a bit more explicit.